src/density_methods.jl

```diff
@@ -0,0 +1,249 @@
+    end
+    ρ
+end
+
+@doc raw"""
+Build a total charge density without spin information from a superposition of atomic
+densities.
+"""
+function atomic_total_density(basis::PlaneWaveBasis{T}, method::AtomicDensity;
+                              coefficients=ones(T, length(basis.model.atoms))) where {T}
+    form_factors = atomic_density_form_factors(basis, method)
+    atomic_density_superposition(basis, form_factors; coefficients)
+end
+
+@doc raw"""
+Build a spin density from a superposition of atomic densities and provided magnetic moments
+(with units ``μ_B``).
+"""
+function atomic_spin_density(basis::PlaneWaveBasis{T}, method::AtomicDensity,
+                             magnetic_moments) where {T}
+    model = basis.model
+    if model.spin_polarization in (:none, :spinless)
+        isempty(magnetic_moments) && return nothing
+        error("Initial magnetic moments can only be used with collinear models.")
+    end
+
+    # If no magnetic moments start with a zero spin density
+    magmoms = Vec3{T}[normalize_magnetic_moment(magmom) for magmom in magnetic_moments]
+    if all(iszero, magmoms)
+        @warn("Returning zero spin density guess, because no initial magnetization has " *
+              "been specified in any of the given elements / atoms. Your SCF will likely " *
+              "not converge to a spin-broken solution.")
+        return zeros(T, basis.fft_size)
+    end
+
+    @assert length(magmoms) == length(basis.model.atoms)
+    coefficients = map(zip(basis.model.atoms, magmoms)) do (atom, magmom)
+        iszero(magmom[1:2]) || error("Non-collinear magnetization not yet implemented")
+        magmom[3] ≤ n_elec_valence(atom) || error(
+            "Magnetic moment $(magmom[3]) too large for element $(atomic_symbol(atom)) " *
+            "with only $(n_elec_valence(atom)) valence electrons."
+        )
+        magmom[3] / n_elec_valence(atom)
+    end
+
+    form_factors = atomic_density_form_factors(basis, method)
+    atomic_density_superposition(basis, form_factors; coefficients)    
+end
+
+@doc raw"""
+Perform an atomic density superposition. The density is constructed in reciprocal space
+using the provided atomic form-factors and coefficients and applying an inverse Fourier
+transform to yield the real-space density.
+"""
+function atomic_density_superposition(basis::PlaneWaveBasis{T},
+                                      form_factors::IdDict{Tuple{Int,T},T};
+                                      coefficients=ones(T, length(basis.model.atoms))
+                                      )::Array{T,3} where {T}
+    model = basis.model
+    G_cart = G_vectors_cart(basis)
+
+    ρ = map(enumerate(G_vectors(basis))) do (iG, G)
+        Gnorm = norm(G_cart[iG])
+        ρ_iG = sum(enumerate(model.atom_groups); init=zero(Complex{T})) do (igroup, group)
+            sum(group) do iatom
+                structure_factor::Complex{T} = cis2pi(-dot(G, model.positions[iatom]))
```

Comment:
I'm doing the functional thing preemptively because, from my vague understanding, it's easier to support on GPU. The annotation may not be necessary, no.